src/mlmc/tool/hdf5.py

```diff
@@ -352,42 +349,37 @@ def scheduled(self):
 
-    def collected(self, chunk_spec=None):
+    def chunks(self, n_samples=None):
+        with h5py.File(self.file_name, 'r') as hdf_file:
+            if 'collected_values' not in hdf_file[self.level_group_path]:
+                raise AttributeError("No collected values in level group ".format(self.level_id))
+            dataset = hdf_file["/".join([self.level_group_path, "collected_values"])]
+
+            if n_samples is not None:
+                yield 0, slice(0, n_samples, 1), int(self.level_id)
+            else:
+                for chunk_id, chunk in enumerate(dataset.iter_chunks()):
+                    yield chunk_id, chunk[0], int(self.level_id)  # slice, level_id
+
+    def collected(self, chunk_slice=None, n_samples=None):
         """
         Read collected data by chunks,
         number of items in chunk is determined by LevelGroup.chunk_size (number of bytes)
-        :param chunk_spec: ChunkSpec instance
+        :param chunk_slice: slice() object
+        :param n_samples: number of samples to retrieve
         :return: np.ndarray
         """
         with h5py.File(self.file_name, 'r') as hdf_file:
             if 'collected_values' not in hdf_file[self.level_group_path]:
                 return None
             dataset = hdf_file["/".join([self.level_group_path, "collected_values"])]
 
-            if chunk_spec is not None:
-                if chunk_spec.n_samples is not None and chunk_spec.n_samples < np.inf:
-                    return dataset[:chunk_spec.n_samples]
-
-                if chunk_spec.chunk_size is not None:
-                    if chunk_spec.chunk_size in self._chunk_size_items:
-                        n_items = self._chunk_size_items[chunk_spec.chunk_size]
-                    else:
-                        first_item = dataset[0]
-                        item_byte_size = first_item.size * first_item.itemsize
-                        n_items = self._chunk_size_items[chunk_spec.chunk_size] = int(np.ceil(chunk_spec.chunk_size / item_byte_size)) \
-                            if int(np.ceil(chunk_spec.chunk_size / item_byte_size)) < len(dataset[()]) else len(dataset[()])
-
-                    self._chunks_info[chunk_spec] = [chunk_spec.chunk_id * n_items, (chunk_spec.chunk_id + 1) * n_items]
-                    return dataset[chunk_spec.chunk_id * n_items: (chunk_spec.chunk_id + 1) * n_items]
-            return dataset[()]
-
-    def get_chunks_info(self, chunk_spec):
-        """
-        The start and end index of a chunk from a whole dataset point of view
-        :param chunk_spec: ChunkSpec instance
-        :return: List[int, int]
-        """
-        return self._chunks_info[chunk_spec]
+            if n_samples is None and chunk_slice is not None:
+                return dataset[chunk_slice]
+            elif n_samples is not None and n_samples < dataset.size:
+                return dataset[:n_samples]
+            else:
+                return dataset[()]
```

Comment:
do you mean dataset[:dataset.size]?
But if the slicing handles over bound slices well returning dataset[chunk_slice] should be fine.

Reply:
I would like to be able to get data without calling the `chunks()` method. And this `else` is the only thing needed to do that.